### PR TITLE
[ci] Run all tests on (self-hosted) GPU

### PR DESCRIPTION
**Pull request by julien-c**

### CODE REVIEW

File: .github/workflows/self-hosted.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    - name: Install dependencies
+      run: |
+        source .env/bin/activate
+        pip install .[mecab,sklearn,tf,torch,testing]
+
+    - name: Are GPU recognized by our DL frameworks
+      run: |
+        source .env/bin/activate
+        python -c "import torch; print(torch.cuda.is_available())"
+        python -c "import tensorflow as tf; print(tf.test.is_built_with_cuda()); print(tf.config.list_physical_devices('GPU'))"
```
Comment (julien-c) [inline .github/workflows/self-hosted.yml]: @n1t0  How do you install cudnn? TF says that it can't use our K80 because it cannot find cudnn.  (cc @LysandreJik and @mfuntowicz)

File: .github/workflows/ubuntu.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    - name: Install dependencies
+      run: |
+        pip install .[sklearn,tf,torch,testing]
+
+    - name: Run all tests on CPU
+      env:
+        OMP_NUM_THREADS: 1
+      run: |
+        python -m pytest -n 8 --dist=loadfile -s -v ./tests/
+
```
Comment (julien-c) [inline .github/workflows/ubuntu.yml]: To benchmark vs. Circle-CI, I've added this one (runs on a GitHub instance)

Seems [almost 3x slower](https://github.com/huggingface/transformers/runs/468406732?check_suite_focus=true) than the machines we have in our Circle CI plan (ofc, it's free vs. we pay Circle CI)

File: .github/workflows/self-hosted.yml (hunk 1 above)
Comment (n1t0) [reply .github/workflows/self-hosted.yml]: I usually follow this: https://docs.nvidia.com/deeplearning/sdk/cudnn-install/index.html#installlinux-tar

Comment (julien-c) [reply .github/workflows/self-hosted.yml]: We installed from the .deb file with @LysandreJik (omg, the Nvidia website is awful)

File: .github/workflows/ubuntu.yml (hunk 2 above)
Comment (mfuntowicz) [reply .github/workflows/ubuntu.yml]: One of the advantage I see using GitHub Actions rather than CircleCI is the ability to dig faster and easier in case on failure. The GA interface is more clear and you can spot directly what was the issue.

File: .github/workflows/ubuntu.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        python-version: 3.7
+    - name: Install dependencies
+      run: |
+        pip install .[sklearn,tf,torch,testing]
+
+    - name: Run all tests on CPU
+      env:
+        OMP_NUM_THREADS: 1
+      run: |
+        python -m pytest -n 8 --dist=loadfile -s -v ./tests/
```
Comment (mfuntowicz) [inline .github/workflows/ubuntu.yml]: FYI with @LysandreJik we suppose TF QA pipeline is failing in some cases because of too many parallel runs leading to OOM on the instance itself.

I've open a branch about this one, launching only 4 parallel works `tf_qa_pipelines_test`.

File: .github/workflows/self-hosted.yml (hunk 1 above)
Comment (mfuntowicz) [reply .github/workflows/self-hosted.yml]: Would it be hard to do the tests with and without CuDNN ? Testing both would cover a broad range of setups.

**Review by mfuntowicz — APPROVED** [review]
Love it !

**Review by patrickvonplaten — APPROVED** [review]

File: .github/workflows/self-hosted.yml (hunk 1 above)
Comment (LysandreJik) [reply .github/workflows/self-hosted.yml]: I believe TensorFlow doesn't recognize the available GPUs when CuDNN is not installed?
Are you talking about PyTorch?

**Review by sshleifer — APPROVED** [review]

File: .github/workflows/self-hosted.yml (hunk 1 above)
Comment (BramVanroy) [reply .github/workflows/self-hosted.yml]: Note that might be of interest: you can't use the [NVidia docker images](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_17.12.html#rel_17.12) with the K80 because they require a minimal compute capability of 6.0 whereas the K80 is rated at 3.7.

**Comment by julien-c** [comment]
**Update:** 
The remaining failing tests:

- Test_doc_samples: multilingual.rst and modeling_flaubert.py (@LysandreJik)
- test_modeling_auto.py::AutoModelTest::test_model_for_pretraining_from_pretrained
- test_modeling_roberta.py::RobertaModelIntegrationTest::test_inference_masked_lm
- test_tokenization_roberta.py::RobertaTokenizationTest::test_sequence_builders
- test_tokenization_xlm_roberta.py::XLMRobertaTokenizationIntegrationTest::test_tokenization_base_hard_symbols

**Comment by patrickvonplaten** [comment]
> * test_tokenization_xlm_roberta.py::XLMRobertaTokenizationIntegrationTest::test_tokenization_base_hard_symbols (@patrickvonplaten [c913eb9](https://github.com/huggingface/transformers/commit/c913eb9c3894b4031dc059d22b42e38a5fcef989))

I think this test currently fails because there is a problem with the xlm_roberta_tokenizer . In general so far there are no tests at all for the xlm_roberta_tokenizer . I can try to add those (will prob need a bit help from @mfuntowicz and @LysandreJik)

**Comment by julien-c** [comment]
Update on my failing unit (integration) test on RoBERTa: the max absolute diff between expected and actual output is `0.004` whereas it used to be under `1e-3` (both CPU and cuda) – should I dive in to why this changed, or should i just lazily bump up to tolerance?

(The same integration test without the maskedLM head still passes with `1e-5` abs diff)

**Comment by LysandreJik** [comment]
Could this be due to the bias? https://github.com/huggingface/transformers/pull/2958

It could have been computed twice when creating the integration test.

**Comment by patrickvonplaten** [comment]
I'm not sure how to exactly get the expected logits from the fairseq **LM** Models. It's very easy to get the last embeddings for verification of the no **LM** model via:

```
import torch
roberta = torch.hub.load('pytorch/fairseq', 'roberta.large')
roberta.eval()  # disable dropout (or leave in train mode to finetune)
last_layer_embeddings = roberta.extract_features(input_ids)   # shape [1, seq_len, 1024]
```

The integration test that were implemented for Roberta LMHeadModels did they correspond to the original weights?

**Comment by sshleifer** [comment]
@patrickvonplaten https://github.com/huggingface/transformers/blob/53ce3854a16ad2a715bc6ac8af3e30c18b5a1d11/tests/test_modeling_roberta.py#L324 is the test (I think)

**Comment by julien-c** [comment]
Ok found a resolution strategy with @LysandreJik, will push fix soon.

**Comment by julien-c** [comment]
Getting closer:
> ===== 6 failed, 939 passed, 32 skipped, 84 warnings in 1508.72s (0:25:08) ======
